cudaextender/src/ungapped_xdrop_kernels.cu

```diff
@@ -0,0 +1,493 @@
+
+    __shared__ int32_t ref_loc[num_warps];
+    __shared__ int32_t query_loc[num_warps];
+    __shared__ int32_t total_score[num_warps];
+    __shared__ int32_t prev_score[num_warps];
+    __shared__ int32_t prev_max_score[num_warps];
+    __shared__ int32_t prev_max_pos[num_warps];
+    __shared__ bool edge_found[num_warps];
+    __shared__ bool xdrop_found[num_warps];
+    __shared__ bool new_max_found[num_warps];
+    __shared__ int32_t left_extent[num_warps];
+    __shared__ int32_t extent[num_warps];
+    __shared__ int32_t tile[num_warps];
+    __shared__ double entropy[num_warps];
+
+    int32_t thread_score;
+    int32_t max_thread_score;
+    int32_t max_pos;
+    int32_t temp_pos;
+    bool xdrop_done;
+    int32_t temp;
+    short count[4];
+    short count_del[4];
+    char r_chr;
+    char q_chr;
+    int32_t ref_pos;
+    int32_t query_pos;
+    int32_t pos_offset;
+
+    __shared__ int32_t sub_mat[nuc2];
+
+    if (thread_id < nuc2)
+    {
+        sub_mat[thread_id] = d_sub_mat[thread_id];
+    }
+    __syncthreads();
+
+    for (int32_t hid0 = block_id * num_warps; hid0 < num_seed_pairs; hid0 += num_warps * num_blocks)
+    {
+        int32_t hid = hid0 + warp_id + start_index;
+
+        if (hid < num_seed_pairs)
+        {
+            if (lane_id == 0)
+            {
+                ref_loc[warp_id]     = d_seed_pairs[hid].target_position_in_read;
+                query_loc[warp_id]   = d_seed_pairs[hid].query_position_in_read;
+                total_score[warp_id] = 0;
+            }
+        }
+        else
+        {
+            if (lane_id == 0)
+            {
+
+                ref_loc[warp_id]     = d_seed_pairs[hid0].target_position_in_read;
+                query_loc[warp_id]   = d_seed_pairs[hid0].query_position_in_read;
+                total_score[warp_id] = 0;
+            }
+        }
+        __syncwarp();
+
+        //////////////////////////////////////////////////////////////////
+        //Right extension
+
+        if (lane_id == 0)
+        {
+            tile[warp_id]           = 0;
+            xdrop_found[warp_id]    = false;
+            edge_found[warp_id]     = false;
+            new_max_found[warp_id]  = false;
+            entropy[warp_id]        = 1.0f;
+            prev_score[warp_id]     = 0;
+            prev_max_score[warp_id] = -1000;
+            prev_max_pos[warp_id]   = 0;
+            extent[warp_id]         = 0;
+        }
+
+        count[0]     = 0;
+        count[1]     = 0;
+        count[2]     = 0;
+        count[3]     = 0;
+        count_del[0] = 0;
+        count_del[1] = 0;
+        count_del[2] = 0;
+        count_del[3] = 0;
+        max_pos      = 0;
+
+        __syncwarp();
+
+        while (!xdrop_found[warp_id] && !edge_found[warp_id])
+        {
+            pos_offset   = lane_id + tile[warp_id];
+            ref_pos      = ref_loc[warp_id] + pos_offset;
+            query_pos    = query_loc[warp_id] + pos_offset;
+            thread_score = 0;
```

Comment:
Again, here and in the following loops: declare the variables as local as possible and add `const` where applicable.

Reply:
Moved variables to local loops. Added const wherever possible.

Reply:
:+1:
Could you move the declaration of `max_pos` and `max_thread_score` just before the if-else statement that initializes them?

Reply:
Done.